### ISSUE TITLE
Good way to know parent of Node in a Visitor?

### ISSUE THREAD

**Issue by s-coimbra21** [ratel-rust/ratel-core#68]
Hey guys, I'm using this crate to build a control flow graph for JS code and I'm wondering how you would implement node parents. I'm reasonably new to Rust and I'm not sure where to start. I'm basing my work on the `rewrite` branch.

**Comment by maciejhirsz** [comment]
Hey @s-coimbra21!

The tree structure only references parent to child, so there is no obvious way to fetch a parent of a node only having a child ref.

Since all AST refs are non-`mut` though, you could implement your own `Context` for the visitor on which you put all references you are recursively diving through onto a stack (using a `Vec` would be fine).

If that doesn't work out, writing your own specialized visitor would be the second best option, it's not that hard and Rust makes sure you branch out into all enum variants for expressions and statements.

**Comment by s-coimbra21** [comment]
The stack sounds like a good idea, I'll just have to make the Visitable implementations to pre-order right?

**Comment by maciejhirsz** [comment]
Yeah, I reckon in order to get parent in the context of a child node you need to swap the lines inside here:

```rust
            Binary(ref binary) => {
                binary.traverse(visitor, ctx);
                visitor.on_binary_expression(binary, self, ctx);
            },
```

I don't know if there are any unforeseen consequences for it, but it's early enough that we can deal with it, so if you do that I'll gladly accept a PR to the rewrite branch.
